### PR TITLE
ci: add PR generator workflow

### PR DESCRIPTION
**Pull request by grant**
# ci: add PR generator GitHub workflow

This PR adds a workflow to this repo that creates PRs when a commit lands to `master` (conditionally if the diff contents is non-zero).

Generated PRs are expected to be reviewed by a human and manually merged.

Fixes: https://github.com/googleapis/google-cloudevents/issues/163

## Summary

**Event:** Push
**Branch:** `master`
**PR commit:**  `feat: run generator (${NOW})"`
**PR contents:** Runs the generator (`./scripts/gen.sh`)

## Example

Screen Shot 2021-02-26 at 16 41 12

Example PR on a fork: https://github.com/grant/google-cloudevents/pull/4

## Detailed Design

Updated the GitHub workflow with the correct information:
- event (push to `master`)
- generator setup (same as compile CI setup)
- generator step (same as diff CI setup)
- git diff conditional (`git status --porcelain`)
- git configs
- unique git branch
- git commit message
- GitHub PR title and body

### Open Questions

- Do we want a different git email/name?
- Do we want to change the git commit message to something different? Is it possible to detect if a change is a `feat` or something else?

## Testing

Tested on a fork with a few PRs.

### CODE REVIEW

**Comment by grant** [comment]
CI will be fixed after merging https://github.com/googleapis/google-cloudevents/pull/173

File: .github/workflows/gen_pr.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: Run Generator and Create Pull Request
+on:
+  push:
+    branches:
+      - "master"
```
Comment (grayside) [inline .github/workflows/gen_pr.yaml:19]: Non-blocking: Filed #177 , I don't think it should block CI, but worth noting each thing added that specifies the `master` branch will make switching more challenging.

File: .github/workflows/gen_pr.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      uses: actions/setup-go@v2
+      with:
+        go-version: '1.14'
+    - name: Run the generators
+      run: ./scripts/gen.sh
+    - name: Commit changelog and manifest files
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      run: |
+        if [[ `git status --porcelain` ]]; then
```
Comment (grayside) [inline .github/workflows/gen_pr.yaml:40]: nit: add comment explaining what condition is being checked.

File: .github/workflows/gen_pr.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          # Changes after running generator
+          git config --global user.email "yoshi-code-bot@google.com"
+          git config --global user.name "Google Bot"
+          git checkout -b "pr_bot-$GITHUB_RUN_ID"
+          git add -A
+          NOW=$(date +"%m-%d-%Y")
+          git commit --message "feat: run generator (${NOW})"
+          git push origin "pr_bot-$GITHUB_RUN_ID"
+          echo "Created commit: $(git rev-parse HEAD)"
+          gh pr create --title "Run the generator (${NOW})" --body "AUTOMATED PR! This PR was created from a recent push to the default branch."
```
Comment (grayside) [inline .github/workflows/gen_pr.yaml:50]: nit: It would be good to add reviewer instructions to the body or a link to reviewer instructions in a README.

**Review by grayside — APPROVED** [review]
Left a couple considerations on the PR but nothing preventing merge.

File: .github/workflows/gen_pr.yaml (hunk 3 above)
Comment (grant) [reply .github/workflows/gen_pr.yaml]: I've added some more instructions in the description here.

We may further refine this as we test this in practice.

**Pull request merged**